In [43]:
import pandas as pd
import numpy as np
import seaborn as sns

In [44]:
car_df = pd.read_csv('/aiffel/data/cars.csv')

In [45]:
brand_df = pd.read_csv('/aiffel/data/brand.csv')

In [46]:
car_df.head()

title  Price  Mileage(miles)  Registration_Year  Previous Owners  \
0     SKODA FABIA   6900           70189               2016              3.0   
1  VAUXHALL CORSA   1495           88585               2008              4.0   
2     HYUNDAI I30    949          137000               2011              NaN   
3      MINI HATCH   2395           96731               2010              5.0   
4  VAUXHALL CORSA   1000           85000               2013              NaN   

  Fuel type  Body type Engine Gearbox  Doors  Seats Emission Class  \
0    Diesel  Hatchback   1.4L  Manual    5.0    5.0         Euro 6   
1    Petrol  Hatchback   1.2L  Manual    3.0    5.0         Euro 4   
2    Petrol  Hatchback   1.4L  Manual    5.0    5.0         Euro 5   
3    Petrol  Hatchback   1.4L  Manual    3.0    4.0         Euro 4   
4    Diesel  Hatchback   1.3L  Manual    5.0    5.0         Euro 5   

  Service history  
0             NaN  
1            Full  
2             NaN  
3            Full  
4             NaN

In [47]:
brand_df.head()

title         country
0     skoda  Czech Republic
1  vauxhall  United Kingdom
2   hyundai     South Korea
3      mini  United Kingdom
4      ford   United States

# cars.csv와 brand.csv의 데이터 합치기

In [48]:
brand_df['title'] = brand_df['title'].str.upper()

In [49]:
brand_df['title']

0             SKODA
1          VAUXHALL
2           HYUNDAI
3              MINI
4              FORD
5             VOLVO
6           PEUGEOT
7               BMW
8           CITROEN
9     MERCEDES-BENZ
10            MAZDA
11             SAAB
12       VOLKSWAGEN
13            HONDA
14               MG
15           TOYOTA
16             SEAT
17           NISSAN
18             ALFA
19          RENAULT
20              KIA
21           PROTON
22             FIAT
23             AUDI
24       MITSUBISHI
25            LEXUS
26             LAND
27        CHEVROLET
28           SUZUKI
29            DACIA
30         DAIHATSU
31             JEEP
32           JAGUAR
33         CHRYSLER
34            ROVER
35               DS
36           DAEWOO
37            DODGE
38          PORSCHE
39           SUBARU
40         INFINITI
41           ABARTH
42            SMART
43           MARCOS
44         MASERATI
45        SSANGYONG
46          LAGONDA
47            ISUZU
Name: title, dtype: object

In [50]:
car_df['brand'] = car_df['title'].apply(lambda x: x.split()[0])

In [51]:
car_df['brand']

0          SKODA
1       VAUXHALL
2        HYUNDAI
3           MINI
4       VAUXHALL
          ...   
3682     RENAULT
3683        AUDI
3684         BMW
3685       HONDA
3686    VAUXHALL
Name: brand, Length: 3687, dtype: object

In [52]:
car_df.head()

title  Price  Mileage(miles)  Registration_Year  Previous Owners  \
0     SKODA FABIA   6900           70189               2016              3.0   
1  VAUXHALL CORSA   1495           88585               2008              4.0   
2     HYUNDAI I30    949          137000               2011              NaN   
3      MINI HATCH   2395           96731               2010              5.0   
4  VAUXHALL CORSA   1000           85000               2013              NaN   

  Fuel type  Body type Engine Gearbox  Doors  Seats Emission Class  \
0    Diesel  Hatchback   1.4L  Manual    5.0    5.0         Euro 6   
1    Petrol  Hatchback   1.2L  Manual    3.0    5.0         Euro 4   
2    Petrol  Hatchback   1.4L  Manual    5.0    5.0         Euro 5   
3    Petrol  Hatchback   1.4L  Manual    3.0    4.0         Euro 4   
4    Diesel  Hatchback   1.3L  Manual    5.0    5.0         Euro 5   

  Service history     brand  
0             NaN     SKODA  
1            Full  VAUXHALL  
2             NaN   HYUNDAI  
3            Full      MINI  
4             NaN  VAUXHALL

In [53]:
merged_df = pd.merge(car_df, brand_df, left_on='brand', right_on='title', how='left')

In [54]:
merged_df

title_x  Price  Mileage(miles)  Registration_Year  \
0        SKODA FABIA   6900           70189               2016   
1     VAUXHALL CORSA   1495           88585               2008   
2        HYUNDAI I30    949          137000               2011   
3         MINI HATCH   2395           96731               2010   
4     VAUXHALL CORSA   1000           85000               2013   
...              ...    ...             ...                ...   
3682  RENAULT MEGANE   1395           76202               2006   
3683         AUDI A4   6990          119000               2012   
3684    BMW 3 SERIES   3995          139000               2013   
3685    HONDA ACCORD   1390          179190               2007   
3686  VAUXHALL CORSA   2000           82160               2013   

      Previous Owners Fuel type  Body type Engine Gearbox  Doors  Seats  \
0                 3.0    Diesel  Hatchback   1.4L  Manual    5.0    5.0   
1                 4.0    Petrol  Hatchback   1.2L  Manual    3.0    5.0   
2                 NaN    Petrol  Hatchback   1.4L  Manual    5.0    5.0   
3                 5.0    Petrol  Hatchback   1.4L  Manual    3.0    4.0   
4                 NaN    Diesel  Hatchback   1.3L  Manual    5.0    5.0   
...               ...       ...        ...    ...     ...    ...    ...   
3682              4.0    Petrol  Hatchback   1.6L  Manual    5.0    5.0   
3683              NaN    Petrol     Saloon   2.0L  Manual    4.0    5.0   
3684              NaN    Diesel     Saloon   2.0L  Manual    4.0    5.0   
3685              NaN    Diesel     Estate   2.2L  Manual    5.0    5.0   
3686              7.0    Petrol  Hatchback   1.2L  Manual    5.0    5.0   

     Emission Class Service history     brand   title_y         country  
0            Euro 6             NaN     SKODA     SKODA  Czech Republic  
1            Euro 4            Full  VAUXHALL  VAUXHALL  United Kingdom  
2            Euro 5             NaN   HYUNDAI   HYUNDAI     South Korea  
3            Euro 4            Full      MINI      MINI  United Kingdom  
4            Euro 5             NaN  VAUXHALL  VAUXHALL  United Kingdom  
...             ...             ...       ...       ...             ...  
3682         Euro 4             NaN   RENAULT   RENAULT          France  
3683         Euro 5             NaN      AUDI      AUDI         Germany  
3684         Euro 5             NaN       BMW       BMW         Germany  
3685         Euro 4            Full     HONDA     HONDA           Japan  
3686         Euro 5             NaN  VAUXHALL  VAUXHALL  United Kingdom  

[3687 rows x 16 columns]

In [55]:
merged_df.drop(columns=['title_y'], inplace=True)

In [56]:
merged_df.head()

title_x  Price  Mileage(miles)  Registration_Year  Previous Owners  \
0     SKODA FABIA   6900           70189               2016              3.0   
1  VAUXHALL CORSA   1495           88585               2008              4.0   
2     HYUNDAI I30    949          137000               2011              NaN   
3      MINI HATCH   2395           96731               2010              5.0   
4  VAUXHALL CORSA   1000           85000               2013              NaN   

  Fuel type  Body type Engine Gearbox  Doors  Seats Emission Class  \
0    Diesel  Hatchback   1.4L  Manual    5.0    5.0         Euro 6   
1    Petrol  Hatchback   1.2L  Manual    3.0    5.0         Euro 4   
2    Petrol  Hatchback   1.4L  Manual    5.0    5.0         Euro 5   
3    Petrol  Hatchback   1.4L  Manual    3.0    4.0         Euro 4   
4    Diesel  Hatchback   1.3L  Manual    5.0    5.0         Euro 5   

  Service history     brand         country  
0             NaN     SKODA  Czech Republic  
1            Full  VAUXHALL  United Kingdom  
2             NaN   HYUNDAI     South Korea  
3            Full      MINI  United Kingdom  
4             NaN  VAUXHALL  United Kingdom

In [57]:
merged_df.rename(columns={'title_x': 'title'}, inplace=True)

In [58]:
merged_df.head()

title  Price  Mileage(miles)  Registration_Year  Previous Owners  \
0     SKODA FABIA   6900           70189               2016              3.0   
1  VAUXHALL CORSA   1495           88585               2008              4.0   
2     HYUNDAI I30    949          137000               2011              NaN   
3      MINI HATCH   2395           96731               2010              5.0   
4  VAUXHALL CORSA   1000           85000               2013              NaN   

  Fuel type  Body type Engine Gearbox  Doors  Seats Emission Class  \
0    Diesel  Hatchback   1.4L  Manual    5.0    5.0         Euro 6   
1    Petrol  Hatchback   1.2L  Manual    3.0    5.0         Euro 4   
2    Petrol  Hatchback   1.4L  Manual    5.0    5.0         Euro 5   
3    Petrol  Hatchback   1.4L  Manual    3.0    4.0         Euro 4   
4    Diesel  Hatchback   1.3L  Manual    5.0    5.0         Euro 5   

  Service history     brand         country  
0             NaN     SKODA  Czech Republic  
1            Full  VAUXHALL  United Kingdom  
2             NaN   HYUNDAI     South Korea  
3            Full      MINI  United Kingdom  
4             NaN  VAUXHALL  United Kingdom

## 데이터 합치기 결과
- car_df와 brand_df로 데이터를 합치기 과정을 실행
- car_df에서는 대문자, brand_df에서는 소문자로 title행이 설정되어있었기 때문에 모두 대문자로 맞춰주었다.
- car_df의 title에서 첫번째 단어를 추출하여 대분자로 변경된 brand_df의 title과 비교하여 알맞게 데이터가 합쳐지도록 해주었다.
- 마지막으로 필요없는 부분을 삭제, 바꿀 부분 rename으로 바꿔주기

# # 카테고리형 변수를 숫자 형태로 변환하기

In [59]:
merged_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3687 entries, 0 to 3686
Data columns (total 15 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   title              3687 non-null   object 
 1   Price              3687 non-null   int64  
 2   Mileage(miles)     3687 non-null   int64  
 3   Registration_Year  3687 non-null   int64  
 4   Previous Owners    2276 non-null   float64
 5   Fuel type          3687 non-null   object 
 6   Body type          3687 non-null   object 
 7   Engine             3642 non-null   object 
 8   Gearbox            3687 non-null   object 
 9   Doors              3662 non-null   float64
 10  Seats              3652 non-null   float64
 11  Emission Class     3600 non-null   object 
 12  Service history    540 non-null    object 
 13  brand              3687 non-null   object 
 14  country            3687 non-null   object 
dtypes: float64(3), int64(3), object(9)
memory usage: 460.9+ KB


In [60]:
merged_df.isna().sum()

title                   0
Price                   0
Mileage(miles)          0
Registration_Year       0
Previous Owners      1411
Fuel type               0
Body type               0
Engine                 45
Gearbox                 0
Doors                  25
Seats                  35
Emission Class         87
Service history      3147
brand                   0
country                 0
dtype: int64

In [61]:
merged_df['Engine'] = merged_df['Engine'].str[:3]

In [62]:
merged_df['Engine']

0       1.4
1       1.2
2       1.4
3       1.4
4       1.3
       ... 
3682    1.6
3683    2.0
3684    2.0
3685    2.2
3686    1.2
Name: Engine, Length: 3687, dtype: object

In [63]:
merged_df['Engine'] = pd.to_numeric(merged_df['Engine'])

In [64]:
merged_df.head()

title  Price  Mileage(miles)  Registration_Year  Previous Owners  \
0     SKODA FABIA   6900           70189               2016              3.0   
1  VAUXHALL CORSA   1495           88585               2008              4.0   
2     HYUNDAI I30    949          137000               2011              NaN   
3      MINI HATCH   2395           96731               2010              5.0   
4  VAUXHALL CORSA   1000           85000               2013              NaN   

  Fuel type  Body type  Engine Gearbox  Doors  Seats Emission Class  \
0    Diesel  Hatchback     1.4  Manual    5.0    5.0         Euro 6   
1    Petrol  Hatchback     1.2  Manual    3.0    5.0         Euro 4   
2    Petrol  Hatchback     1.4  Manual    5.0    5.0         Euro 5   
3    Petrol  Hatchback     1.4  Manual    3.0    4.0         Euro 4   
4    Diesel  Hatchback     1.3  Manual    5.0    5.0         Euro 5   

  Service history     brand         country  
0             NaN     SKODA  Czech Republic  
1            Full  VAUXHALL  United Kingdom  
2             NaN   HYUNDAI     South Korea  
3            Full      MINI  United Kingdom  
4             NaN  VAUXHALL  United Kingdom

In [65]:
merged_df['Emission Class'] = pd.to_numeric(merged_df['Emission Class'].str[4:])

In [66]:
merged_df.head()

title  Price  Mileage(miles)  Registration_Year  Previous Owners  \
0     SKODA FABIA   6900           70189               2016              3.0   
1  VAUXHALL CORSA   1495           88585               2008              4.0   
2     HYUNDAI I30    949          137000               2011              NaN   
3      MINI HATCH   2395           96731               2010              5.0   
4  VAUXHALL CORSA   1000           85000               2013              NaN   

  Fuel type  Body type  Engine Gearbox  Doors  Seats  Emission Class  \
0    Diesel  Hatchback     1.4  Manual    5.0    5.0             6.0   
1    Petrol  Hatchback     1.2  Manual    3.0    5.0             4.0   
2    Petrol  Hatchback     1.4  Manual    5.0    5.0             5.0   
3    Petrol  Hatchback     1.4  Manual    3.0    4.0             4.0   
4    Diesel  Hatchback     1.3  Manual    5.0    5.0             5.0   

  Service history     brand         country  
0             NaN     SKODA  Czech Republic  
1            Full  VAUXHALL  United Kingdom  
2             NaN   HYUNDAI     South Korea  
3            Full      MINI  United Kingdom  
4             NaN  VAUXHALL  United Kingdom

### 카테고리형 변수를 숫자 형태로 변환하기
- 일단 먼저 제공한 힌트를 처리
- Engine에서 L 빼기
- Emission Class에서 Euro 빼기
- to_numeric으로 숫자형으로 변경

# 결측치 처리하기

In [67]:
merged_df[merged_df.duplicated()]

title  Price  Mileage(miles)  Registration_Year  \
12              SKODA FABIA   6900           70189               2016   
20               CITROEN C4   1949          114000               2013   
25        VAUXHALL INSIGNIA   2950           79000               2011   
39    MERCEDES-BENZ A CLASS  15900           62402               2019   
52           VAUXHALL CORSA  10698           22782               2020   
...                     ...    ...             ...                ...   
3662             MINI HATCH  11999           33156               2018   
3663         VAUXHALL ASTRA   4999           60139               2015   
3674   MITSUBISHI OUTLANDER   1489          145000               2007   
3675         VAUXHALL ASTRA   4999           60139               2015   
3676         VAUXHALL ASTRA   4999           60139               2015   

      Previous Owners Fuel type  Body type  Engine    Gearbox  Doors  Seats  \
12                3.0    Diesel  Hatchback     1.4     Manual    5.0    5.0   
20                NaN    Diesel  Hatchback     1.6     Manual    5.0    5.0   
25                4.0    Petrol  Hatchback     1.8     Manual    5.0    5.0   
39                1.0    Petrol  Hatchback     1.3  Automatic    5.0    5.0   
52                NaN    Petrol  Hatchback     1.2     Manual    5.0    5.0   
...               ...       ...        ...     ...        ...    ...    ...   
3662              1.0    Petrol  Hatchback     1.5     Manual    5.0    5.0   
3663              2.0    Petrol  Hatchback     1.6     Manual    5.0    5.0   
3674              NaN    Diesel        SUV     2.0     Manual    5.0    7.0   
3675              2.0    Petrol  Hatchback     1.6     Manual    5.0    5.0   
3676              2.0    Petrol  Hatchback     1.6     Manual    5.0    5.0   

      Emission Class Service history          brand         country  
12               6.0             NaN          SKODA  Czech Republic  
20               5.0             NaN        CITROEN          France  
25               5.0             NaN       VAUXHALL  United Kingdom  
39               6.0             NaN  MERCEDES-BENZ         Germany  
52               6.0             NaN       VAUXHALL  United Kingdom  
...              ...             ...            ...             ...  
3662             6.0            Full           MINI  United Kingdom  
3663             6.0             NaN       VAUXHALL  United Kingdom  
3674             4.0             NaN     MITSUBISHI           Japan  
3675             6.0             NaN       VAUXHALL  United Kingdom  
3676             6.0             NaN       VAUXHALL  United Kingdom  

[828 rows x 15 columns]

In [68]:
merged_df = merged_df.drop_duplicates()

In [69]:
merged_df

title  Price  Mileage(miles)  Registration_Year  \
0        SKODA FABIA   6900           70189               2016   
1     VAUXHALL CORSA   1495           88585               2008   
2        HYUNDAI I30    949          137000               2011   
3         MINI HATCH   2395           96731               2010   
4     VAUXHALL CORSA   1000           85000               2013   
...              ...    ...             ...                ...   
3682  RENAULT MEGANE   1395           76202               2006   
3683         AUDI A4   6990          119000               2012   
3684    BMW 3 SERIES   3995          139000               2013   
3685    HONDA ACCORD   1390          179190               2007   
3686  VAUXHALL CORSA   2000           82160               2013   

      Previous Owners Fuel type  Body type  Engine Gearbox  Doors  Seats  \
0                 3.0    Diesel  Hatchback     1.4  Manual    5.0    5.0   
1                 4.0    Petrol  Hatchback     1.2  Manual    3.0    5.0   
2                 NaN    Petrol  Hatchback     1.4  Manual    5.0    5.0   
3                 5.0    Petrol  Hatchback     1.4  Manual    3.0    4.0   
4                 NaN    Diesel  Hatchback     1.3  Manual    5.0    5.0   
...               ...       ...        ...     ...     ...    ...    ...   
3682              4.0    Petrol  Hatchback     1.6  Manual    5.0    5.0   
3683              NaN    Petrol     Saloon     2.0  Manual    4.0    5.0   
3684              NaN    Diesel     Saloon     2.0  Manual    4.0    5.0   
3685              NaN    Diesel     Estate     2.2  Manual    5.0    5.0   
3686              7.0    Petrol  Hatchback     1.2  Manual    5.0    5.0   

      Emission Class Service history     brand         country  
0                6.0             NaN     SKODA  Czech Republic  
1                4.0            Full  VAUXHALL  United Kingdom  
2                5.0             NaN   HYUNDAI     South Korea  
3                4.0            Full      MINI  United Kingdom  
4                5.0             NaN  VAUXHALL  United Kingdom  
...              ...             ...       ...             ...  
3682             4.0             NaN   RENAULT          France  
3683             5.0             NaN      AUDI         Germany  
3684             5.0             NaN       BMW         Germany  
3685             4.0            Full     HONDA           Japan  
3686             5.0             NaN  VAUXHALL  United Kingdom  

[2859 rows x 15 columns]

In [71]:
merged_df['Service history'] = merged_df['Service history'].fillna('Unknown')

/tmp/ipykernel_55/1662461919.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  merged_df['Service history'] = merged_df['Service history'].fillna('Unknown')


In [73]:
merged_df.isna().sum()

title                  0
Price                  0
Mileage(miles)         0
Registration_Year      0
Previous Owners      983
Fuel type              0
Body type              0
Engine                17
Gearbox                0
Doors                 25
Seats                 35
Emission Class        59
Service history        0
brand                  0
country                0
dtype: int64

In [76]:
merged_df['Doors'].fillna(merged_df['Doors'].median(), inplace = True)

/opt/conda/lib/python3.9/site-packages/pandas/core/generic.py:6392: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return self._update_inplace(result)


In [78]:
merged_df['Seats'].fillna(merged_df['Seats'].median(), inplace = True)

/opt/conda/lib/python3.9/site-packages/pandas/core/generic.py:6392: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return self._update_inplace(result)


In [79]:
merged_df['Engine'].fillna(merged_df['Engine'].median(), inplace = True)

/opt/conda/lib/python3.9/site-packages/pandas/core/generic.py:6392: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return self._update_inplace(result)


In [80]:
merged_df['Emission Class'].fillna(merged_df['Emission Class'].median(), inplace = True)

/opt/conda/lib/python3.9/site-packages/pandas/core/generic.py:6392: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return self._update_inplace(result)


In [82]:
merged_df.dropna(subset = ['Previous Owners'], inplace = True)

/opt/conda/lib/python3.9/site-packages/pandas/util/_decorators.py:311: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return func(*args, **kwargs)


In [84]:
merged_df.isna().sum()

title                0
Price                0
Mileage(miles)       0
Registration_Year    0
Previous Owners      0
Fuel type            0
Body type            0
Engine               0
Gearbox              0
Doors                0
Seats                0
Emission Class       0
Service history      0
brand                0
country              0
dtype: int64

### 결측치 결과
- 데이터의 양이 그리 크지 않기때문에 최대한 데이터를 가지고 가는 쪽으로 해서 진행하려고 했음
- 하지만 일단 중복값들은 제거해야 수월하게 진행할 수 있을 것 같아 중복값 제거
- 그리고 제공된 힌트를 가지고 먼저 해결,  Service history의 NaN값을 Unknown으로 채워주기(fillna() 사용)
- 나머지 NaN값이 존재하는 Engine, Doors, Seats, Emission Class, Previous Owners 결측치 해결
- Engine, Doors, Seats, Emission Class는 fillna()로 median으로 중앙값으로 변경
- Previous Owners는 고민을 많이 했다. 이전 소유자가 NaN이면 뭘까 생각하다가...그냥 없애버렸다,, 얘도 중앙값으로 하려다가 고민의 시간이 너무 오래되어 일단 제거하였음(나머지처럼 fillna()로 중앙값으로 변경하는 것은 나중에 해보기로 )
- 어찌됐든 결측치는 모두 제거를 완료하였음.

# Scaler 활용하기, PCA 주성분 분석

In [96]:
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import RobustScaler
from sklearn.preprocessing import MinMaxScaler

### StandardScaler

In [97]:
ss = StandardScaler()

In [98]:
ss_cols = ['Price', 'Mileage(miles)', 'Registration_Year', 'Previous Owners', 'Engine', 'Doors', 'Seats', 'Emission Class']

In [99]:
ss_df = ss.fit_transform(merged_df[ss_cols])

In [100]:
pd.DataFrame(ss_df, columns = ss_cols)

Price  Mileage(miles)  Registration_Year  Previous Owners    Engine  \
0     0.322196       -0.325563           0.880151         0.076292 -0.463419   
1    -0.947873        0.108153          -0.763820         0.723909 -0.873019   
2    -0.736391        0.300208          -0.352827         1.371527 -0.463419   
3    -1.111184        0.947741          -0.969316         0.076292  0.765379   
4    -0.993929        0.070784          -0.763820         1.371527 -0.053820   
...        ...             ...                ...              ...       ...   
1871 -0.652973        0.377280          -1.585805         0.076292  4.246974   
1872 -0.717592        1.699859          -2.818784        -0.571326  1.174978   
1873 -0.900877        0.471587          -0.558324         0.076292 -0.053820   
1874 -0.971371       -0.183796          -1.174813         0.723909 -0.053820   
1875 -0.829208       -0.043327           0.263662         2.666762 -0.873019   

         Doors     Seats  Emission Class  
0     0.647437  0.147701        1.269054  
1    -1.441284  0.147701       -1.035637  
2    -1.441284 -1.521499       -1.035637  
3    -1.441284 -1.521499       -1.035637  
4     0.647437  0.147701       -1.035637  
...        ...       ...             ...  
1871  0.647437  0.147701       -2.187982  
1872 -2.485644 -1.521499       -3.340328  
1873  0.647437  0.147701       -1.035637  
1874  0.647437  0.147701       -1.035637  
1875  0.647437  0.147701        0.116709  

[1876 rows x 8 columns]

#### 기존 데이터의 정보 70% 이상을 가지는 수준에서 최소한의 주성분 추출 (PCA)

In [102]:
from sklearn.decomposition import PCA

In [111]:
pca = PCA(n_components = 0.7)

In [112]:
pca.fit(ss_df)

PCA(n_components=0.7)

In [113]:
pd.DataFrame(pca.fit_transform(ss_df))

0         1         2
0    -1.503838 -0.324262 -0.222088
1     1.671419  0.754423 -1.068227
2     1.858165  1.867019 -0.555930
3     2.342484  1.560915  0.483300
4     1.707094 -0.680034 -0.738637
...        ...       ...       ...
1871  3.048952 -1.165115  2.947107
1872  4.501970  1.971874  0.907724
1873  1.236544 -0.758629 -0.555786
1874  1.593314 -0.614483 -0.824385
1875  0.805836 -0.522907 -1.187669

[1876 rows x 3 columns]

In [114]:
(pca.explained_variance_ratio_).sum() 

0.7139656397086738

###  RobustScaler

In [117]:
rs = RobustScaler()

In [118]:
rs_df = rs.fit_transform(merged_df[ss_cols])

In [119]:
pd.DataFrame(rs_df, columns = ss_cols)

Price  Mileage(miles)  Registration_Year  Previous Owners    Engine  \
0     0.605112       -0.254500           0.666667              0.0 -0.285714   
1    -0.522692        0.140771          -0.666667              0.5 -0.571429   
2    -0.334898        0.315802          -0.333333              1.0 -0.285714   
3    -0.667710        0.905936          -0.833333              0.0  0.571429   
4    -0.563589        0.106714          -0.666667              1.0  0.000000   
...        ...             ...                ...              ...       ...   
1871 -0.260824        0.386042          -1.333333              0.0  3.000000   
1872 -0.318206        1.591386          -2.333333             -0.5  0.857143   
1873 -0.480960        0.471989          -0.500000              0.0  0.000000   
1874 -0.543558       -0.125300          -1.000000              0.5  0.000000   
1875 -0.417319        0.002718           0.166667              2.0 -0.571429   

      Doors  Seats  Emission Class  
0       0.0    0.0             0.5  
1      -2.0    0.0            -0.5  
2      -2.0   -1.0            -0.5  
3      -2.0   -1.0            -0.5  
4       0.0    0.0            -0.5  
...     ...    ...             ...  
1871    0.0    0.0            -1.0  
1872   -3.0   -1.0            -1.5  
1873    0.0    0.0            -0.5  
1874    0.0    0.0            -0.5  
1875    0.0    0.0             0.0  

[1876 rows x 8 columns]

In [120]:
pca.fit(rs_df)

PCA(n_components=0.7)

In [121]:
pd.DataFrame(pca.fit_transform(rs_df))

0         1         2
0    -0.967427  0.406644 -0.139460
1     1.344715 -0.975906 -0.684371
2     1.431405 -1.327709 -0.326039
3     1.910791 -1.083623  0.388197
4     1.124173  0.741452 -0.746934
...        ...       ...       ...
1871  1.689110  0.838552  1.252398
1872  3.143015 -1.747601  0.908536
1873  0.807978  0.830729 -0.362570
1874  1.005866  0.662007 -0.897731
1875  0.712715  0.673701 -0.954889

[1876 rows x 3 columns]

In [122]:
(pca.explained_variance_ratio_).sum() 

0.7329184987434334

### MinMaxScaler

In [123]:
mm = MinMaxScaler()

In [124]:
mm_df = mm.fit_transform(merged_df[ss_cols])

In [125]:
pd.DataFrame(mm_df, columns = ss_cols)

Price  Mileage(miles)  Registration_Year  Previous Owners    Engine  \
0     0.194030        0.063223           0.862745            0.250  0.109091   
1     0.032687        0.079794           0.705882            0.375  0.072727   
2     0.059552        0.087132           0.745098            0.500  0.109091   
3     0.011940        0.111873           0.686275            0.250  0.218182   
4     0.026836        0.078366           0.705882            0.500  0.145455   
...        ...             ...                ...              ...       ...   
1871  0.070149        0.090077           0.627451            0.250  0.527273   
1872  0.061940        0.140611           0.509804            0.125  0.254545   
1873  0.038657        0.093680           0.725490            0.250  0.145455   
1874  0.029701        0.068639           0.666667            0.375  0.145455   
1875  0.047761        0.074006           0.803922            0.750  0.072727   

         Doors  Seats  Emission Class  
0     1.000000    0.6             1.0  
1     0.333333    0.6             0.6  
2     0.333333    0.4             0.6  
3     0.333333    0.4             0.6  
4     1.000000    0.6             0.6  
...        ...    ...             ...  
1871  1.000000    0.6             0.4  
1872  0.000000    0.4             0.2  
1873  1.000000    0.6             0.6  
1874  1.000000    0.6             0.6  
1875  1.000000    0.6             0.8  

[1876 rows x 8 columns]

In [126]:
pca.fit(mm_df)

PCA(n_components=0.7)

In [127]:
pd.DataFrame(pca.fit_transform(mm_df))

0         1
0    -0.252206 -0.094313
1     0.504292  0.105032
2     0.559340  0.141091
3     0.529354  0.039567
4    -0.087609  0.389771
...        ...       ...
1871 -0.074335  0.393034
1872  0.911504  0.126522
1873 -0.138151  0.233837
1874 -0.107023  0.326005
1875 -0.096897  0.381735

[1876 rows x 2 columns]

In [128]:
(pca.explained_variance_ratio_).sum() 

0.7702467319673387

# 국가별 총 브랜드 개수

In [116]:
merged_df.groupby('country')['brand'].count().sort_values()

country
Malaysia            1
Romania            10
Sweden             27
Spain              31
Czech Republic     32
Italy              62
South Korea        98
United States     226
France            264
Japan             317
United Kingdom    360
Germany           448
Name: brand, dtype: int64

### 국가별 총 브랜드 개수 결과
- Malaysia가 1개로 가장 적다
- Germany가 448개로 가장 많다
- 한국은 98개의 브랜드가 있다.(처음 알았다..저렇게 많았구낭)

# 회고
지난 시간에 늦게 와서 진도를 끝까지 마치지 못하고 주말동안 공부를 해보았는데 엄청 잘 이해가 되진 않았다..그래서 생각보다 이번 프로젝트를 하면서 복습의 시간을 많이 가졌고 이것저것 많이 찾아보기도 했지만 역시 저번 노드를 다시 찾아보는 게 제일 도움이 되었다.
이번 프로젝트를 진행하면서 내가 너무 복잡하게 생각해서 코드에 뭐를 덕지덕지 붙이다보니 데이터가 이상해지는 것 같다..일단 문제를 단편적으로 이해하고 뼈대를 세우는게 좋을 것 같다. 데이터양이 적으면 최대한 데이터를 보존해서 가져가는게 좋다해서 그렇게 하고 싶었는데 내 사고력의 문제로 다수의 데이터를 삭제해버렸다..ㅜㅠ물론 삭제하는게 나쁜건 아니지만 생각한대로 진행되지 않음에 대한 후회가 좀 있다. 후에 시간이 날 때 이 프로젝트를 리팩토링 해봐야겠다.
부족했던 부분으로는 scaler가 뒤돌면 까먹는 느낌이라 좀 더 레퍼런스도 찾아보고 정리하는 시간을 가져야겠다.